### PR TITLE
Question Queue

### PR DESCRIPTION
**Pull request by shiquan-lim**
- Question Queue
  - Students may post questions to an event-specific question forum
  - Questions are sorted by number of peer upvotes
  - Asker may mark questions as answered by a particular response from peers, educators, and assistants
  - Asker may alert educators to especially pertinent questions
- Event joining bugfix
  - events are now joinable again
  - Issues remain with updateRanking() and accessing eventParticipants(). I believe the error lies in a similar location to the joinEvent() bug, namely in services.js and an issue with the security rules where auth nodes were not readable

### CODE REVIEW

File: database/security-rules.json [hunk 1]
```diff
@@ -6,7 +6,8 @@
       "publicIds": {
         "$publicId": {
           ".validate": "newData.isString() && newData.val().matches(/^[-:a-zA-Z0-9]+$/) && $publicId.matches(/^[-a-z0-9]{3,20}$/) && ($publicId == null || newData.parent().parent().parent().child('auth').child('publicIds').child($publicId).val() == newData.val() && newData.parent().parent().parent().child('auth').child('usedPublicIds').child($publicId).val() == true && newData.parent().parent().parent().child('auth').child('users').child(newData.val()).child('publicId').val() == $publicId)",
-          ".write": "auth != null && auth.uid == newData.val() && (newData.val() != null && data.val() == null)"
+          ".write": "auth != null && auth.uid == newData.val() && (newData.val() != null && data.val() == null)",
+          ".read": "auth != null"
```
Comment (dinoboff) [inline database/security-rules.json:10]: I am not sure we should.

It's fine for new database; I think the new UID format is unique to each DB and I am assuming it's opaque. For firebase DB imported from the legacy Firebase console, old UID are not unique and should be consider private data (you can track user from different DB). It probably expose the google user UID.

File: database/security-rules.json [hunk 2]
```diff
@@ -160,7 +182,7 @@
         "$userActions": {
           ".write": "newData.hasChildren(['publicId','action','timestamp'])",
           "publicId": {
-            ".validate": "newData.isString() && root.child('auth/publicIds').child(newData.val()).val() === auth.uid"
+            ".validate": " true || newData.isString() && root.child('auth/publicIds').child(newData.val()).val() === auth.uid"
```
Comment (dinoboff) [inline database/security-rules.json]: I understand why for rapid development, we might overlook or loosen the rules, but we should document somehow what was the issue, maybe with a skipped test.

**Comment by dinoboff** [comment]
Can we please stop reformatting services.js?

File: database/security-rules.json (hunk 1 above)
Comment (dinoboff) [inline database/security-rules.json]: I see... it was to fixed joining event. Joining event should be using auth/usedPublicIds. It's now fixed.

Comment (dinoboff) [reply database/security-rules.json:10]: Ok, it was a fix to the buggy event.join method. It has been fix.

Comment (shiquan-lim) [reply database/security-rules.json:10]: Yes thank you very much

File: database/security-rules.json (hunk 2 above)
Comment (shiquan-lim) [reply database/security-rules.json]: @dinoboff This was for the purpose of team activities. In an team, only the leader submits an answer, but we want to log all users as having collaboratively submitted something. Hence this security rule of blocking writes for user actions not containing one's own publicID was an issue. We will see what work arounds we can apply, but tentatively we shall leave it as it is

**Comment by shiquan-lim** [comment]
@dinoboff apologies for the services reformat. We will try not to do it again

File: database/security-rules.json (hunk 2 above)
Comment (SingaporeClouds) [reply database/security-rules.json]: I don't think we should let any user log a userAction for any other user. userActions are intended to be the logs for individual user's actions. If you want to have the logging user include the publicIds of their team members, that might be better than relaxing the security rules.

Comment (shiquan-lim) [reply database/security-rules.json]: Fair enough. I can fix and push that up now. And then revert the security rules to its previous state without the forced true

Comment (shiquan-lim) [reply database/security-rules.json]: Changes pushed. Team submissions are now only logged by the leader as such:

image

**Comment by shiquan-lim** [comment]
Please do not merge. I have yet to update the naming of question queue to questions

**Pull request merged**